host/integration_test.go

```diff
@@ -1591,25 +1606,25 @@ func (s *integrationSuite) TestActivityHeartbeatTimeouts() {
 
 	atHandler := func(execution *workflow.WorkflowExecution, activityType *workflow.ActivityType,
 		activityID string, input []byte, taskToken []byte) ([]byte, bool, error) {
-		s.Equal(id, *execution.WorkflowId)
-		s.Equal(activityName, *activityType.Name)
-		timeoutType := string(input)
-		switch timeoutType {
-		case "Heartbeat":
-			s.logger.Info("Starting hearbeat activity.")
-			go func() {
-				for i := 0; i < 6; i++ {
-					s.logger.Infof("Heartbeating for activity: %s, count: %d", activityID, i)
-					_, err := s.engine.RecordActivityTaskHeartbeat(createContext(), &workflow.RecordActivityTaskHeartbeatRequest{
-						TaskToken: taskToken, Details: []byte(strconv.Itoa(i))})
-					s.Nil(err)
-					time.Sleep(1 * time.Second)
+		s.logger.Infof("Starting heartbeat activity. ID: %v", activityID)
+		for i := 0; i < 10; i++ {
+			if !workflowComplete {
+				s.logger.Infof("Heartbeating for activity: %s, count: %d", activityID, i)
+				_, err := s.engine.RecordActivityTaskHeartbeat(createContext(), &workflow.RecordActivityTaskHeartbeatRequest{
+					TaskToken: taskToken, Details: []byte(strconv.Itoa(i))})
+				if err != nil {
+					s.logger.Errorf("Activity heartbeat failed.  ID: %v, Progress: %v, Error: %v", activityID, i, err)
 				}
-				s.logger.Info("End Heartbeating.")
-			}()
-			s.logger.Info("Sleeping hearbeat activity.")
-			time.Sleep(12 * time.Second)
+
+				secondsToSleep := rand.Intn(3)
+				s.logger.Infof("Activity ID '%v' sleeping for: %v seconds", activityID, secondsToSleep)
+				time.Sleep(time.Duration(secondsToSleep) * time.Second)
+			}
 		}
+		s.logger.Infof("End Heartbeating. ID: %v", activityID)
+
+		s.logger.Infof("Sleeping activity before completion. ID: %v", activityID)
+		time.Sleep(5 * time.Second)
```

Comment:
with the sleep in the loop, this test would run for up to 35s? that seems very long for one test.

Reply:
It will be roughly 25 seconds for this test.  Can't think of another way to test this for now.  Let's work on running the test in parallel to reduce the time.